feat(dogfood): implement `ExportGenesis`

MaxMustermann2 opened the pull request:
GenesisState defines the dogfood module's genesis state. Note that, as always,
`genesis` is a misnomer. Ideally, this state can be exported at any point in
time (or height), and reimported elsewhere where it will be the new genesis
potentially at a non-zero height. In other words, it is the entire, current,
state of the module.

The state consists of the following items.
- `ValSet`, the currently active validator set
- `OptOutExpiries`, which refers to a list of epochs and a corresponding list of
operator account addresses whose opt-outs will expire at the end of said epoch.
- `ConsensusAddrsToPrune` which contains a similar list of epochs and a
corresponding list of consensus addresses to be pruned from the operator
module at the end of said epoch.
- `UndelegationMaturities` which contains another list of epochs and a
corresponding list of undelegation record keys to mature at the end of
said epoch.
- `LastTotalPower` which represents the last recorded power of the currently
active validator set.

For convenience:
- The `ValSet` consists of the `bytes32` key and voting power.
- `OptOutExpiries` contain the bech32 `AccAddress`
- `ConsensusAddrsToPrune` contain the bech32 `ConsAddress`
- `UndelegationMaturities` contain the hex encoded record key

Other changes:
- Operator queries with `chain-id` supplied report whether the process of opting out is still under process (`OptingOut`). The consensus key is stored until the opt-out epoch + unbonding duration is complete, and hence, it is returned even when the operator is in the process of opting out. To avoid confusion, an `OptingOut` boolean flag is returned.
- The dogfood module now correctly allows querying of the validator by its consensus address.
- BREAKING: The process to opt into the dogfood-AVS has now changed. Instead of requiring that a consensus key be supplied separately, it must be supplied as the second argument during opt-in. Similarly, opting out is now a one-command process, subsuming the `init-cons-key-removal` command into itself.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced support for new types and structures in the genesis state, including `EpochToOperatorAddrs`, `EpochToConsensusAddrs`, and `EpochToUndelegationRecordKeys`.
  - Added new functions to handle opt-out, pruning, and undelegation processes.

- **Enhancements**
  - Updated Go version from `1.21.9` to `1.21.11` across various workflows and configuration files.
  - Improved initialization and validation logic in genesis state-related functions.
  - Enhanced Dockerfile base image to use Go `1.21.11`.

- **Bug Fixes**
  - Fixed issues related to validator set handling and public key validation in genesis state.

- **Refactor**
  - Modified conditional logic in GitHub workflows to improve reliability and maintainability.
  - Adjusted import statements and function signatures to align with new dependencies and structures.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MaxMustermann2 committed:
feat(dogfood): re-scaffold genesis

In preparation for the `ExportGenesis` implementation, re-scaffold the
genesis state of the dogfood module.

MaxMustermann2 committed:
fix(dogfood): remove deprecated keys

MaxMustermann2 committed:
refactor: use new dogfood genesis

MaxMustermann2 committed:
refactor(dogfood): - validatorupdate from gen

MaxMustermann2 committed:
refactor(dogfood): update ImportGenesis

...for new genesis structure

MaxMustermann2 committed:
refactor(dogfood): simplify gen state names

MaxMustermann2 committed:
feat(dogfood): implement ExportGenesis

Partially resolves #54

MaxMustermann2 committed:
chore: lint

MaxMustermann2 committed:
proto(dogfood): remove unused import

MaxMustermann2 committed:
fix(local): compat with new dogfood genesis

MaxMustermann2 committed:
chore: lint proto file and gen

MaxMustermann2 committed:
fix(test): use correct struct member name

MaxMustermann2 committed:
chore: update go version to 1.21.11

MaxMustermann2 committed:
fix(ci): disable buggy workflows

till they can be fixed

MaxMustermann2 committed:
chore: lint

bwhour reviewed (approved):
it looks good.

cloud8little commented:
Trying to export the latest height, query the validator information before stopping the node, there are two active validators with power 5000,1001. after exporting, there is only one validator in the `val_set`. 

```
exocored q tendermint-validator-set
block_height: "159864"
total: "2"
validators:
- address: exovalcons18z3p42xn8pjk338upvzp794h02wh7p4t7jj9jx
  proposer_priority: "1268"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: 8PaRnlIsW5fbLIJVv/dD+d/d162fw3ywwWcLSA0PmRQ=
  voting_power: "5000"
- address: exovalcons1fllfq6nj5tc9wu8a82dtqd8axfxva48wzna4pa
  proposer_priority: "-1267"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: XLtFCK0/nB1xExSXEhH5kaxRte3aIXSGaBfWSeNOtpE=
  voting_power: "1001"
```

```
    "dogfood": {
      "consensus_addrs_to_prune": [],
      "last_total_power": "6001",
      "opt_out_expiries": [],
      "params": {
        "asset_ids": [
          "0xdac17f958d2ee523a2206206994597c13d831ec7_0x65"
        ],
        "epoch_identifier": "hour",
        "epochs_until_unbonded": 7,
        "historical_entries": 10000,
        "max_validators": 100
      },
      "undelegation_maturities": [],
      "val_set": [
        {
          "power": "5000",
          "public_key": "0xf0f6919e522c5b97db2c8255bff743f9dfddd7ad9fc37cb0c1670b480d0f9914"
        }
      ]
    }
```

MaxMustermann2 committed:
fix(dogfood): don't stop iterating in export

The return value should be `false` to continue

MaxMustermann2 commented:
> Trying to export the latest height, query the validator information before stopping the node, there are two active validators with power 5000,1001. after exporting, there is only one validator in the `val_set`.

Yes, my bad. I only tested with one operator via `./local_node.sh`. Fixed in the latest commit 97723ed by continuing the loop after the first iteration.

MaxMustermann2 commented on local_node.sh:
```diff
@@ -115,10 +115,11 @@ if [[ $overwrite == "y" || $overwrite == "Y" ]]; then
 	# x/dogfood
-	jq '.app_state["dogfood"]["initial_val_set"][0]["public_key"]="0xf0f6919e522c5b97db2c8255bff743f9dfddd7ad9fc37cb0c1670b480d0f9914"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
-	jq '.app_state["dogfood"]["initial_val_set"][0]["power"]="5000"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
+	jq '.app_state["dogfood"]["val_set"][0]["public_key"]="0xf0f6919e522c5b97db2c8255bff743f9dfddd7ad9fc37cb0c1670b480d0f9914"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
+	jq '.app_state["dogfood"]["val_set"][0]["power"]="5000"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
 	# change the epoch to an hour when starting a local node, which facilitates the testing.
 	jq '.app_state["dogfood"]["params"]["epoch_identifier"]="hour"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
+	jq '.app_state["dogfood"]["last_total_power"]="5000"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
```
This is a testing script and not a production script. It is used to quickly produce a single-validator (to begin with) localnet.

MaxMustermann2 commented on x/dogfood/keeper/impl_sdk.go:
```diff
@@ -167,7 +164,7 @@ func (k Keeper) ApplyAndReturnValidatorSetUpdates(
 // the SDK's gov module and by our oracle module.
 func (k Keeper) IterateBondedValidatorsByPower(
-	ctx sdk.Context, f func(int64, stakingtypes.ValidatorI) bool,
+	ctx sdk.Context, f func(int64, stakingtypes.ValidatorI) (stop bool),
```
The error handling is a weaker version of the one done by Cosmos SDK, which panics if it is unable to find an operator (`mustGetValidator`). I would prefer that we don't panic unless there is something catastrophically wrong.

MaxMustermann2 replied:
Added logs in e8731d0f95f092f5193c46c9c6a3a00226a6af29

MaxMustermann2 committed:
fix(doc): add comment about exported validators

MaxMustermann2 commented on x/dogfood/types/genesis_test.go:
```diff
@@ -84,64 +111,534 @@
+			expPass: true,
+		},
+		{
+			name: "duplicate epoch in pruning",
+			genState: &types.GenesisState{
+				Params: params,
+				ValSet: []types.GenesisValidator{
+					{
+						PublicKey: sharedKey,
+						Power:     5,
+					},
+				},
+				LastTotalPower: math.NewInt(5),
+				ConsensusAddrsToPrune: []types.EpochToConsensusAddrs{
+					{
+						Epoch: 2,
+						ConsAddrs: []string{
+							consAddr1.String(),
+						},
+					},
+					{
+						Epoch: 2,
+						ConsAddrs: []string{
+							consAddr1.String(),
+						},
+					},
+				},
+			},
+			expPass:  false,
+			expError: "duplicate epoch",
```
This is a test verifying the exact behaviour to which you're pointing. I am not sure what is the actionable comment here.

MaxMustermann2 commented on proto/exocore/dogfood/v1/genesis.proto:
```diff
@@ -2,28 +2,113 @@
 
-  // initial_val_set is the initial validator set.
-  repeated GenesisValidator initial_val_set = 2
+  // val_set is the initial validator set. it onyl represents the active
+  // validators.
+  repeated GenesisValidator val_set = 2
```
Not in production; ignoring.

MaxMustermann2 commented on x/dogfood/keeper/genesis.go:
```diff
@@ -69,6 +109,26 @@ func (k Keeper) InitGenesis(
 	genesis.Params = k.GetDogfoodParams(ctx)
-	// TODO(mm)
-	return genesis
+	validators := []types.GenesisValidator{}
+	k.IterateBondedValidatorsByPower(ctx, func(_ int64, val stakingtypes.ValidatorI) bool {
+		// #nosec G703 // already validated
+		pubKey, _ := val.ConsPubKey()
+		// #nosec G703 // already validated
+		convKey, _ := cryptocodec.ToTmPubKeyInterface(pubKey)
+		validators = append(validators,
+			types.GenesisValidator{
+				PublicKey: hexutil.Encode(convKey.Bytes()),
+				Power:     val.GetConsensusPower(sdk.DefaultPowerReduction),
+			},
+		)
+		return false /* stop */
+	})
+	return types.NewGenesis(
+		k.GetDogfoodParams(ctx),
+		validators,
+		k.GetAllOptOutsToFinish(ctx),
+		k.GetAllConsAddrsToPrune(ctx),
+		k.GetAllUndelegationsToMature(ctx),
+		k.GetLastTotalPower(ctx),
+	)
```
These factors are validated on the fly; for example, the dogfood module will never save more than the max number of validators to disk anyway. Adding more checks here, to do the same thing, will be overkill.

MaxMustermann2 commented on x/dogfood/keeper/unbonding.go:
```diff
@@ -87,6 +88,33 @@ func (k Keeper) setUndelegationsToMature(
 	store.Set(key, val)
 }
 
+// GetAllUndelegationsToMature gets a list of epochs and the corresponding undelegation record
+// keys which are scheduled to mature at the end of that epoch. It is ordered, first by the
+// epoch and then by the record key's bytes.
+func (k Keeper) GetAllUndelegationsToMature(ctx sdk.Context) []types.EpochToUndelegationRecordKeys {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{types.OptOutsToFinishBytePrefix})
+	defer iterator.Close()
+
+	res := []types.EpochToUndelegationRecordKeys{}
+
+	for ; iterator.Valid(); iterator.Next() {
+		epoch, _ := types.SafeUint64ToInt64(sdk.BigEndianToUint64(iterator.Key()[1:]))
+		var recordKeys types.UndelegationRecordKeys
+		k.cdc.MustUnmarshal(iterator.Value(), &recordKeys)
+		subRes := []string{}
+		for _, recordKey := range recordKeys.GetList() {
+			subRes = append(subRes, hexutil.Encode(recordKey))
+		}
+		res = append(res, types.EpochToUndelegationRecordKeys{
+			Epoch:                  epoch,
+			UndelegationRecordKeys: subRes,
+		})
+	}
+
+	return res
+}
```
Please suggest an alternative but keep in mind that this function is called only during `ExportGenesis` when the node is offline.

MaxMustermann2 replied:
I don't understand this.

The iterator is processing exactly one record per iteration; that is what an iterator typically does. So how will batching help?

The record keys have a common pattern, that is, they are stored as a key value pair with the key being a prefix + big-endian epoch (the "pattern") and the value being the identifier of the undelegation record. This is as direct as it gets.

MaxMustermann2 committed:
doc(dogfood): log iteration failure

If such an iteration failure happens during export, a pipe to `jq` will
fail and inform the user of the failure.

MaxMustermann2 commented on x/dogfood/types/genesis_test.go:
```diff
@@ -22,31 +27,49 @@ func TestGenesisTestSuite(t *testing.T) {
 
 func (suite *GenesisTestSuite) TestValidateGenesis() {
 	sharedKey := hexutil.Encode(ed25519.GenPrivKey().PubKey().Bytes())
+	operator1 := sdk.AccAddress(testutiltx.GenerateAddress().Bytes())
+	consAddr1 := sdk.ConsAddress(operator1)
+	recordKey := hexutil.Encode(
+		delegationtypes.GetUndelegationRecordKey(
+			1000, // block height
+			1,    // layer zero nonce
+			common.BytesToHash([]byte("tx hash")).Hex(),
+			operator1.String(),
+		),
+	)
```
Out of scope for this PR.

MaxMustermann2 replied:
Added comment to `GetUndelegationRecordKey` that the caller must validate the inputs to the function. They are currently (stateless) validated everywhere this function is called, except here, which is a unit test.

MaxMustermann2 committed:
docs(delegation): validation of record key params

MaxMustermann2 commented on app/export.go:
```diff
@@ -75,14 +76,9 @@ func (app *ExocoreApp) ExportAppStateAndValidators(
 	return servertypes.ExportedApp{
 		AppState:        appState,
-		Validators:      validators,
+		Validators:      []tmtypes.GenesisValidator{}, // exported directly in dogfood
```
Added comments

MaxMustermann2 commented on app/test_helpers.go:
```diff
@@ -274,6 +275,8 @@ func GenesisStateWithValSet(app *ExocoreApp, genesisState simapp.GenesisState,
 			},
 		},
+		[]dogfoodtypes.EpochToOperatorAddrs{}, []dogfoodtypes.EpochToConsensusAddrs{},
+		[]dogfoodtypes.EpochToUndelegationRecordKeys{}, math.NewInt(1),
```
Added comments

MaxMustermann2 commented on testutil/utils.go:
```diff
@@ -240,6 +240,10 @@ func (suite *BaseTestSuite) SetupWithGenesisValSet(genAccs []authtypes.GenesisAc
+		[]dogfoodtypes.EpochToOperatorAddrs{},
+		[]dogfoodtypes.EpochToConsensusAddrs{},
+		[]dogfoodtypes.EpochToUndelegationRecordKeys{},
+		math.NewInt(2),
```
Added comments

cloud8little committed:
Merge branch 'develop' into fix/dogfood-export-gen

MaxMustermann2 commented:
~The exported genesis is erroneous. I am checking.~

This has now been fixed.

```json
{
  "consensus_addrs_to_prune": [
    {
      "cons_addrs": [
        "exovalcons1vwjcr2h5tfz0nxqnf0aw8dfslzz6nn7m00fjls"
      ],
      "epoch": "12"
    }
  ],
  "last_total_power": "5000",
  "opt_out_expiries": [
    {
      "epoch": "12",
      "operator_acc_addrs": [
        "exo1vwjcr2h5tfz0nxqnf0aw8dfslzz6nn7m4qg6kt"
      ]
    }
  ],
  "params": {
    "asset_ids": [
      "0xdac17f958d2ee523a2206206994597c13d831ec7_0x65"
    ],
    "epoch_identifier": "minute",
    "epochs_until_unbonded": 2,
    "historical_entries": 10000,
    "max_validators": 100
  },
  "undelegation_maturities": [
    {
      "epoch": "12",
      "undelegation_record_keys": [
        "0x63a581aaf45a44f998134bfae3b530f885a9cfdb"
      ]
    }
  ],
  "val_set": [
    {
      "power": "5000",
      "public_key": "0xf0f6919e522c5b97db2c8255bff743f9dfddd7ad9fc37cb0c1670b480d0f9914"
    }
  ]
}
```

MaxMustermann2 committed:
Merge develop

MaxMustermann2 committed:
fix(app): register operator hooks for dogfood

MaxMustermann2 committed:
feat(operator): return `OptingOut` status

Whenever a query for consensus keys or addresses is answered, it should
specify whether the operator is currently in the process of opting out.
This is because the data is retained until the opt out is complete, even
though the operator may not be actively validating. The purpose of the
data retention, as always, is to slash the operator (if required), based
on the key.

MaxMustermann2 committed:
refactor(operator): rationalize opt-in/out

When an operator opts into an AVS, they should be required to supply the
consensus (or other) key to opt-in. Similarly, when they opt-out, the
key should automatically be removed.

This commit implements that functionality, however, it is restricted to
`ctx.ChainID()`. For an overbroad implementation, an AVS registry from
AVS address -> key type should be implemented.

MaxMustermann2 committed:
refactor(delegation): verbose panic

MaxMustermann2 committed:
fix(dogfood): query validator by cons addr not hex

MaxMustermann2 committed:
refactor(dogfood): move opt func to correct loc

MaxMustermann2 committed:
chore(x/operator): lint go

MaxMustermann2 commented on x/delegation/keeper/genesis.go:
```diff
@@ -44,7 +45,7 @@ func (k Keeper) InitGenesis(
 				}
 				if err := k.delegateTo(ctx, delegationParams, false); err != nil {
-					panic(err)
+					panic(errorsmod.Wrap(err, "failed to delegate to operator"))
```
The `InitGenesis` function signature is fixed. I cannot alter it. 

Since this is a problematic error - one that is observed during genesis - it should panic.

MaxMustermann2 commented on x/operator/keeper/msg_server.go:
```diff
@@ -20,23 +21,55 @@ func (k *Keeper) RegisterOperator(ctx context.Context, req *types.RegisterOperat
-func (k Keeper) OptIntoAVS(ctx context.Context, req *types.OptIntoAVSReq) (*types.OptIntoAVSResponse, error) {
-	c := sdk.UnwrapSDKContext(ctx)
+func (k Keeper) OptIntoAVS(ctx context.Context, req *types.OptIntoAVSReq) (res *types.OptIntoAVSResponse, err error) {
+	uncachedCtx := sdk.UnwrapSDKContext(ctx)
+	// only write if both calls succeed
+	c, writeFunc := uncachedCtx.CacheContext()
+	defer func() {
+		if err == nil {
+			writeFunc()
+		}
+	}()
+	// TODO: use some form of an AVS to key-type registry here, possibly from within the AVS module to determine
+	// if a key is required and that it is appropriately supplied.
+	if req.AvsAddress == c.ChainID() {
+		if len(req.PublicKey) == 0 {
+			return nil, errorsmod.Wrap(types.ErrInvalidPubKey, "a key is required but was not supplied")
+		}
+	} else {
+		if len(req.PublicKey) > 0 {
+			return nil, errorsmod.Wrap(types.ErrInvalidPubKey, "a key is not required but was supplied")
+		}
+	}
 	// #nosec G703 // already validated
 	accAddr, _ := sdk.AccAddressFromBech32(req.FromAddress)
-	err := k.OptIn(c, accAddr, req.AvsAddress)
+	err = k.OptIn(c, accAddr, req.AvsAddress)
 	if err != nil {
 		return nil, err
 	}
+	if len(req.PublicKey) > 0 {
+		// we have to validate just the key; we previously validated that ctx.ChainID() == req.AvsAddress
+		keyObj, _ := types.ValidateConsensusKeyJSON(req.PublicKey)
+		if err := k.SetOperatorConsKeyForChainID(c, accAddr, req.AvsAddress, keyObj); err != nil {
+			return nil, err
+		}
+	}
```
This is the `TODO` note in there. We need to create an AVS to key-type registry for identifying and validating that the key is supplied where it needs to be (and not supplied where a key is not required). Please make an issue about this for later.

MaxMustermann2 commented on x/operator/keeper/msg_server.go:
```diff
@@ -20,23 +21,55 @@
-func (k Keeper) OptOutOfAVS(ctx context.Context, req *types.OptOutOfAVSReq) (*types.OptOutOfAVSResponse, error) {
-	c := sdk.UnwrapSDKContext(ctx)
+func (k Keeper) OptOutOfAVS(ctx context.Context, req *types.OptOutOfAVSReq) (res *types.OptOutOfAVSResponse, err error) {
+	uncachedCtx := sdk.UnwrapSDKContext(ctx)
+	// only write if both calls succeed
+	c, writeFunc := uncachedCtx.CacheContext()
+	defer func() {
+		if err == nil {
+			writeFunc()
+		}
+	}()
 	// #nosec G703 // already validated
 	accAddr, _ := sdk.AccAddressFromBech32(req.FromAddress)
-	err := k.OptOut(c, accAddr, req.AvsAddress)
+	err = k.OptOut(c, accAddr, req.AvsAddress)
```
This error is already validated in `ValidateBasic` for the `req` type.

MaxMustermann2 committed:
fix(operator): wrap and return error

MaxMustermann2 commented on x/operator/types/consensus_keys.go:
```diff
@@ -79,3 +79,18 @@
+		return nil, errorsmod.Wrap(ErrParameterInvalid, "invalid public key type")
+	} else if res, err := StringToPubKey(keyString); err != nil {
+		return nil, errorsmod.Wrap(err, "invalid public key")
+	} else {
```
Fixed in d0c2e6e833b0db8f6433927e6d22f09f8ccaf394

MaxMustermann2 commented on x/operator/types/msg.go:
```diff
@@ -72,12 +69,8 @@ func (m *SetConsKeyReq) ValidateBasic() error {
-	} else if _, err := StringToPubKey(keyString); err != nil {
-		return errorsmod.Wrap(err, "invalid public key")
+	if _, err := ValidateConsensusKeyJSON(m.PublicKey); err != nil {
+		return err
```
Fixed in add0753ede5df98839ddbc9b8f3938093eedc8cf

cloud8little commented:
Test passed (https://github.com/ExocoreNetwork/exocore/pull/95/commits/add0753ede5df98839ddbc9b8f3938093eedc8cf), set epoch as `minute` and **epochs_until_unbonded** as 5, after sending opt-out-avs transaction at epoch 176. and export genesis file like: 

```json
    "dogfood": {
      "consensus_addrs_to_prune": [
        {
          "cons_addrs": [
            "exovalcons1u8yx7e02wr2yvs5zk3zhvehhekh4vzxde43yra"
          ],
          "epoch": "180"
        }
      ],
      "last_total_power": "5500",
      "opt_out_expiries": [
        {
          "epoch": "180",
          "operator_acc_addrs": [
            "exo1u8yx7e02wr2yvs5zk3zhvehhekh4vzxdr6sv2x"
          ]
        }
      ],
      "params": {
        "asset_ids": [
          "0xdac17f958d2ee523a2206206994597c13d831ec7_0x65"
        ],
        "epoch_identifier": "minute",
        "epochs_until_unbonded": 5,
        "historical_entries": 10000,
        "max_validators": 100
      },
      "undelegation_maturities": [
        {
          "epoch": "180",
          "undelegation_record_keys": [
            "0xe1c86f65ea70d4464282b4457666f7cdaf5608cd"
          ]
        }
      ],
      "val_set": [
        {
          "power": "5000",
          "public_key": "0xf0f6919e522c5b97db2c8255bff743f9dfddd7ad9fc37cb0c1670b480d0f9914"
        },
        {
          "power": "500",
          "public_key": "0x5cbb4508ad3f9c1d711314971211f991ac51b5edda2174866817d649e34eb691"
        }
      ]
    },
```

after epoch 180, export genesis file like:  

  ```json
  "dogfood": {
      "consensus_addrs_to_prune": [],
      "last_total_power": "5000",
      "opt_out_expiries": [],
      "params": {
        "asset_ids": [
          "0xdac17f958d2ee523a2206206994597c13d831ec7_0x65"
        ],
        "epoch_identifier": "minute",
        "epochs_until_unbonded": 5,
        "historical_entries": 10000,
        "max_validators": 100
      },
      "undelegation_maturities": [],
      "val_set": [
        {
          "power": "5000",
          "public_key": "0xf0f6919e522c5b97db2c8255bff743f9dfddd7ad9fc37cb0c1670b480d0f9914"
        }
      ]
    },
```
